src/Microsoft.AspNet.Routing/Template/TemplateBinder.cs

```diff
@@ -220,14 +220,21 @@ public string BindValues(IDictionary<string, object> acceptedValues)
                             // we won't necessarily add it to the URI we generate.
                             if (!context.Buffer(converted))
                             {
-                                return null;
+                                    return null;
                             }
                         }
                         else
                         {
                             if (!context.Accept(converted))
                             {
-                                return null;
+                                if (part.IsOptional && segment.Parts[j - 1].IsOptionalSeperator)
+                                {
+                                    context.Remove(segment.Parts[j - 1].Text);
+                                }
```

Comment:
No, in case like this {action}/{name?} -- the context accepts it and we are never here. For other cases, the parser makes sure that in a composite segment the optional parameter is always preceded by optional separator. I'll still add the check.